### PR TITLE
perf: cache schema types

### PR DESCRIPTION
**Pull request by nikpivkin**
This PR adds caching of schema types.

Fixes: https://github.com/open-policy-agent/opa/issues/6969

### CODE REVIEW

**Comment by anderseknert** [comment]
Looks like a small enough change to be able to land before the next release goes out this week. @ashutosh-narkar and @johanfylling's call.

File: ast/check.go [hunk 1]
```diff
@@ -196,20 +199,42 @@ func (tc *typeChecker) checkClosures(env *TypeEnv, expr *Expr) Errors {
...
 			tc.err([]*Error{err})
 			continue
 		}
-		if ref == nil && refType == nil {
+
+		if refType == nil {
 			continue
 		}
+
+		ref := schemaAnnot.Path
```
Comment (ashutosh-narkar) [inline ast/check.go:237]: We can move this above

```go
if refType == nil {
	continue
}
```

and keep the original check

```go
if ref == nil && refType == nil {
    continue
}
```

**Review by ashutosh-narkar — COMMENTED** [review]
This change looks good. It would be helpful if you could add a test case.

**Comment by anderseknert** [comment]
The type checker can't be accessed concurrently, right? 😅

**Comment by ashutosh-narkar** [comment]
> The type checker can't be accessed concurrently, right? 😅

There are existing map fields and I don't see a mutex so maybe not. But I haven't checked the code.

File: ast/check.go (hunk 1 above)
Comment (nikpivkin) [reply ast/check.go]: Fixed

**Comment by nikpivkin** [comment]
@ashutosh-narkar How can you test caching if it's an internal implementation? Need to add performance tests?

**Comment by nikpivkin** [comment]
The tests on darwin failed, but from the logs I don't see why.

**Comment by ashutosh-narkar** [comment]
> @ashutosh-narkar How can you test caching if it's an internal implementation? Need to add performance tests?

If we could add some unit tests for `checkRule` for example that exercise the cache, that would be great.

**Comment by simar7** [comment]
> The tests on darwin failed, but from the logs I don't see why.

@ashutosh-narkar @anderseknert do you know why the tests are failing? We don't see results in the CI log output stating anything is failing.

**Comment by nikpivkin** [comment]
Hi @ashutosh-narkar . The tests already cover all cache use cases. How can a test look like, without breaking encapsulation and explicitly checking for the presence of values in the cache?

**Review by ashutosh-narkar — APPROVED** [review]

**Pull request merged**